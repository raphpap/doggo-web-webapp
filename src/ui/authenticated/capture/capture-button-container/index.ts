export * from './capture-button-container';
export {default} from './capture-button-container';
