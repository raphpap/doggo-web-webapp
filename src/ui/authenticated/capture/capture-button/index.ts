export * from './capture-button';
export {default} from './capture-button';
