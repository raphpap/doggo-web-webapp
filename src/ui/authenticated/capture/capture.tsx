// Vendor
import React from 'react';
import Webcam from 'react-webcam';
import {compose} from 'recompose';

// Assets
import logo from 'doggo/assets/images/capture-logo.png';

// Context
import {
  withApplicationContext,
  WithApplicationContextProps
} from 'doggo/context';

// Shared Components
import Message from 'doggo/ui/@components/message';
import Modal from 'doggo/ui/@components/modal';
import PrettyButton from 'doggo/ui/@components/pretty-button';

// Components
import CaptureButtonContainer from './capture-button-container';
import Form from './form';
import HiddenWebcamContainer from './hidden-webcam-container';
import WebcamContainer from './webcam-container';

// Constants
// const videoConstraints = {
//   facingMode: 'environment',
//   height: 720,
//   width: 1280
// };

// Types
interface Props {}
type EnhancedProps = Props & WithApplicationContextProps;

const enhance = compose<EnhancedProps, Props>(withApplicationContext);

interface State {
  capturedImageSrc: string | null;
  isWebcamActive: boolean;
  isHiddenWebcamActive: boolean;
}

export class Capture extends React.Component<EnhancedProps, State> {
  public readonly state: State = {
    capturedImageSrc: null,
    isHiddenWebcamActive: false,
    isWebcamActive: false
  };

  private hiddenWebcam = React.createRef() as any;

  private webcamTimer?: number;
  private hiddenWebcamTimer?: number;

  public componentWillUnmount() {
    window.clearTimeout(this.webcamTimer);
    window.clearTimeout(this.hiddenWebcamTimer);
  }

  public render() {
    const {capturedImageSrc, isHiddenWebcamActive, isWebcamActive} = this.state;
    const disableCapture = !isHiddenWebcamActive || !isWebcamActive;

    return (
      <>
        <Message>
          Find a Doggo and press Capture to make him part of your team!
        </Message>

        <div>
          <WebcamContainer>
            <Webcam
              /* width={340} */
              /* height={260} */
              width={680}
              height={520}
              audio={false}
              /* videoConstraints={videoConstraints} */
              onUserMedia={this.onWebcamActivated}
            />

            {!disableCapture && !capturedImageSrc && (
              <CaptureButtonContainer>
                <PrettyButton logo={logo} onClick={this.onCaptureClicked}>
                  Capture
                </PrettyButton>
              </CaptureButtonContainer>
            )}
          </WebcamContainer>

          <HiddenWebcamContainer>
            <Webcam
              width={680}
              height={520}
              audio={false}
              /* videoConstraints={videoConstraints} */
              onUserMedia={this.onHiddenWebcamActivated}
              ref={this.setHiddenWebcamRef}
            />
          </HiddenWebcamContainer>
        </div>

        <Modal isOpen={!!capturedImageSrc} onClose={this.closeModal}>
          {capturedImageSrc && (
            <Form
              imageSrc={capturedImageSrc}
              onClose={this.closeModal}
              onSubmit={this.handleSubmit}
            />
          )}
        </Modal>
      </>
    );
  }

  private setHiddenWebcamRef = (webcam: any) => {
    this.hiddenWebcam = webcam;
  };

  private onWebcamActivated = () => {
    this.webcamTimer = window.setTimeout(() => {
      this.setState({isWebcamActive: true});
    }, 3000);
  };

  private onHiddenWebcamActivated = () => {
    this.hiddenWebcam.getScreenshot();

    this.hiddenWebcamTimer = window.setTimeout(() => {
      this.setState({isHiddenWebcamActive: true});
    }, 3000);
  };

  private onCaptureClicked = () => {
    const imageSrc = this.hiddenWebcam.getScreenshot() as string;
    this.setState({capturedImageSrc: imageSrc});
  };

  private closeModal = () => {
    this.setState({capturedImageSrc: null});
  };

  private handleSubmit = (cardName: string, imageSrc: string) => {
    const {capture} = this.props.context.actions;
    capture(cardName, imageSrc);
    this.closeModal();
  };
}

export default enhance(Capture);
