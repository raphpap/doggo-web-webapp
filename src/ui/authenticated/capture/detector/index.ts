export * from './detector';
export {default} from './detector';
