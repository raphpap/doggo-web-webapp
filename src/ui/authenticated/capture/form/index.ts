export * from './form';
export {default} from './form';
