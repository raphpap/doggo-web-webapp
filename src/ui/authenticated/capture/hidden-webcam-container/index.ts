export * from './hidden-webcam-container';
export {default} from './hidden-webcam-container';
