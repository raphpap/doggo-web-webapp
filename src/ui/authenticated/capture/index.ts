export * from './capture';
export {default} from './capture';
