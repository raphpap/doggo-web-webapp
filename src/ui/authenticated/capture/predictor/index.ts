export * from './predictor';
export {default} from './predictor';
