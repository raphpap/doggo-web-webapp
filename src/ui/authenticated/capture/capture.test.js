// Vendor
import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

// Assets
import logo from 'doggo/assets/images/capture-logo.png';

// Components
import {Capture} from './capture';
import Form from './form';

// Shared Components
import Modal from 'doggo/ui/@components/modal';
import PrettyButton from 'doggo/ui/@components/pretty-button';

// Mocks
import {mockCard} from 'doggo/test-mocks';

// Mocks
import {mockLoggedIn} from 'doggo/test-mocks';

describe('<Capture />', () => {
  let wrapper;
  let context;

  beforeEach(() => {
    context = mockLoggedIn();
    wrapper = shallow(<Capture context={context} />);
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('does not shows the form modal before a capture', () => {
    const modal = wrapper.find(Modal);
    expect(modal.props().isOpen).toBe(false);

    const form = wrapper.find(Form);
    expect(form).toHaveLength(0);
  });

  it('shows the form modal after a capture', () => {
    wrapper.setState({capturedImageSrc: 'image_mocked'});

    const modal = wrapper.find(Modal);
    expect(modal.props().isOpen).toBe(true);

    const form = wrapper.find(Form);
    expect(form).toHaveLength(1);
  });

  it('does not render the capture button before the webcam is ready', () => {
    wrapper.setState({
      isHiddenWebcamActive: true,
      isWebcamActive: false
    });

    const prettyButton = wrapper.find(PrettyButton);
    expect(prettyButton).toHaveLength(0);
  });

  it('does not render the capture button before the hidden webcam is ready', () => {
    wrapper.setState({
      isHiddenWebcamActive: false,
      isWebcamActive: true
    });

    const prettyButton = wrapper.find(PrettyButton);
    expect(prettyButton).toHaveLength(0);
  });

  it('renders the capture button when both webcams are ready', () => {
    wrapper.setState({
      isHiddenWebcamActive: true,
      isWebcamActive: true
    });

    const prettyButton = wrapper.find(PrettyButton);
    expect(prettyButton).toHaveLength(1);
    expect(prettyButton.props().logo).toBe(logo);
  });
});
