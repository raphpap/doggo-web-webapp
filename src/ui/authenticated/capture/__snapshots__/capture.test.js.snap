// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Capture /> renders correctly 1`] = `
<Fragment>
  <Message>
    Find a Doggo and press Capture to make him part of your team!
  </Message>
  <div>
    <WebcamContainer>
      <Webcam
        audio={false}
        className=""
        height={520}
        onUserMedia={[Function]}
        onUserMediaError={[Function]}
        screenshotFormat="image/webp"
        screenshotQuality={0.92}
        width={680}
      />
    </WebcamContainer>
    <HiddenWebcamContainer>
      <Webcam
        audio={false}
        className=""
        height={520}
        onUserMedia={[Function]}
        onUserMediaError={[Function]}
        screenshotFormat="image/webp"
        screenshotQuality={0.92}
        width={680}
      />
    </HiddenWebcamContainer>
  </div>
  <ModalComponent
    isOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;
